### PR TITLE
crush: improve straw2 algorithm's readability

### PR DESCRIPTION
**Pull request by yaozongyou**
Signed-off-by: Yao Zongyou <yaozongyou@vip.qq.com>

### CODE REVIEW

**Comment by yaozongyou** [comment]
@liewegas may you please have a look at this? try to improve straw2's readability.
For learning straw2's internal, i have read some material, for example:
https://www.spinics.net/lists/ceph-devel/msg21635.html
the description in this mail is not specifc~~
I have written some small code to verify the algorithm:
https://gist.github.com/yaozongyou/53e3d7722c1f1d6dcddb4675206ee65a
the output is:
DistributionTest DrawStraw:
0 12601
1 12489
2 12537
3 12439
4 12634
5 12372
6 12400
7 12528

DistributionTest DrawStrawEx:
0 23532
1 11757
2 11683
3 11746
4 11830
5 11790
6 11765
7 5897

DistributionTest DrawStrawCeph:
0 23568
1 11718
2 11841
3 11744
4 11872
5 11607
6 11744
7 5906

so in effect, the straw2 has the same result with exponential distribution.

File: src/crush/mapper.c [hunk 1]
```diff
@@ -319,40 +326,53 @@ static inline __s32 *get_choose_arg_ids(const struct crush_bucket_straw2 *bucket
...
-			ln = crush_ln(u) - 0x1000000000000ll;
-
-			/*
-			 * divide by 16.16 fixed-point weight.  note
-			 * that the ln value is negative, so a larger
-			 * weight means a larger (less negative) value
-			 * for draw.
-			 */
-			draw = div64_s64(ln, weights[i]);
+                        draw = generate_exponential_distribution(bucket->h.hash, x, ids[i], r, weights[i]);
```
Comment (tchaikov) [inline src/crush/mapper.c]: this change just extracts the well documented 4-lines code block into a separate function. i doubt it improves the readability.

Comment (yaozongyou) [reply src/crush/mapper.c]: the original document only tells how to compute the natural log or how to divide, 
but here, the key point is exponential distribution, how to generate exponential distribution based on random(), which is evenly distribution.

File: src/crush/mapper.c [hunk 2]
```diff
@@ -293,10 +293,17 @@ static __u64 crush_ln(unsigned int xin)
...
+ * Suppose we have two osds: osd.0 and osd.1, with weight 8 and 4 respectively, It means:
+ *   a). For osd.0, the time interval between each io request apply to exponential distribution 
+ *       with lamba equals 8
+ *   b). For osd.1, the time interval between each io request apply to exponential distribution 
+ *       with lamba equals 4
+ *   c). If we apply to each osd's exponential random variable, then the total pgs on each osd
+ *       is proportional to its weight.
  *
- * http://en.wikipedia.org/wiki/Exponential_distribution#Distribution_of_the_minimum_of_exponential_random_variables
+ * for reference, see:
```
Comment (liewegas) [inline src/crush/mapper.c]: maybe worth keeping an additional note about the properties of the min?

Comment (yaozongyou) [reply src/crush/mapper.c]: I think it has little to do with `the properties of the min`,  each osd draws straw by following its own pseudo-random exponential distribution sequence. Osds share nothing between each other.

Comment (liewegas) [reply src/crush/mapper.c]: the whole reason we use this distribution is that if you pick the min (max, if negative) after dividing by the weight then the resulting distribution *of placements* is proportional to the weight.  that isn't the case if you use a uniform distribution for the per-osd random raw... it's the whole reason we bother with ln() etc.

Comment (yaozongyou) [reply src/crush/mapper.c]: the formula coincidence with exponential distribution~~
see:  http://preshing.com/20111007/how-to-generate-random-timings-for-a-poisson-process/
and my test:
https://gist.github.com/yaozongyou/53e3d7722c1f1d6dcddb4675206ee65a.

the process can be considered as each osd draws straws by following its own pseudo-random exponential distribution sequence. and at last, we compare each osd' straw, and the longest straw wins. And by the exponential distribution math principle, the min random means happening faster, so the min exponential distribution random variable, the longer straw.

File: src/crush/mapper.c [hunk 3]
```diff
@@ -319,40 +326,51 @@ static inline __s32 *get_choose_arg_ids(const struct crush_bucket_straw2 *bucket
...
+   */
+  __s64 ln = crush_ln(u) - 0x1000000000000ll;
+
+  /*
+   * divide by 16.16 fixed-point weight.  note
+   * that the ln value is negative, so a larger
+   * weight means a larger (less negative) value
+   * for draw.
+   */
+  return div64_s64(ln, weight);
```
Comment (liewegas) [inline src/crush/mapper.c]: oops, one formatting nit: can you use 8 chars or tab here?  this file is formatted in kernel C style.  thanks!

Comment (yaozongyou) [reply src/crush/mapper.c]: changed

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-02-01_09:17:08-rados-wip-kefu-testing-2018-02-01-1258-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-02-02_05:30:30-rados-wip-kefu-testing-2018-02-01-1258-distro-basic-mira/

**Pull request merged**